src/main/java/com/uid2/admin/vertx/service/SiteService.java

```diff
@@ -196,4 +208,75 @@ private void handleSiteEnable(RoutingContext rc) {
             rc.fail(500, e);
         }
     }
+
+    private void handleSiteDomains(RoutingContext rc) {
+        try {
+            // refresh manually
+            siteProvider.loadContent();
+
+            final Site existingSite = RequestUtil.getSite(rc, "id", siteProvider);
+            if (existingSite == null) {
+                return;
+            }
+
+            JsonObject body = rc.body().asJsonObject();
+            JsonArray domainNamesJa = body.getJsonArray("domain_names");
+            if(domainNamesJa == null) {
+                ResponseUtil.error(rc, 400, "required parameters: domain_names");
+                return;
+            }
+            List<String> domainNames = domainNamesJa.stream().map(String::valueOf).collect(Collectors.toList());
+
+            List<String> normalizedDomainNames = new ArrayList<>();
+            for(String domain : domainNames) {
+                try {
+                    String tld = getTopLevelDomainName(domain);
+                    normalizedDomainNames.add(tld);
+                } catch (Exception e) {
+                    ResponseUtil.error(rc, 400, "invalid domain name: " + domain);
+                    return;
+                }
+            }
+
+            boolean containsDuplicates = normalizedDomainNames.stream().distinct().count() < normalizedDomainNames.size();
+            if (containsDuplicates) {
+                ResponseUtil.error(rc, 400, "duplicate domain_names not permitted");
+                return;
+            }
+
+            existingSite.setDomainNames(new HashSet<>(normalizedDomainNames));
+
+            final List<Site> sites = this.siteProvider.getAllSites()
+                    .stream().sorted(Comparator.comparingInt(Site::getId))
+                    .collect(Collectors.toList());
+
+            storeWriter.upload(sites, null);
+
+            rc.response().end(jsonWriter.writeValueAsString(existingSite));
+        } catch (Exception e) {
+            ResponseUtil.errorInternal(rc, "set site domain_names failed", e);
+        }
+    }
+
+    public static String getTopLevelDomainName(String origin) throws MalformedURLException {
```

Comment:
maybe we should raise a new PR in UID2 shared lib with this function added so both operator and admin service can use it and we use the same single method to get domain name to be consistent?

Reply:
the admin ui can expect/accpet domains that are passed as top-level domains that are not valid urls. like considering asfd.net valid instead of expecting it to be input as https://asdf.net. So the logic is a bit different I think.  If operator is fine to use this same behavior then may as well move it, otherwise I think the method small enough to just keep here.

Reply:
ok, here is a ticket for this. Let me know if the description looks correct. [UID2-1659](https://atlassian.thetradedesk.com/jira/browse/UID2-1659)

Reply:
looks good thanks! putting it on top of backlog for now. Please work on this when we merge admin/operator back to master